lib/mllp/client.ex

```diff
@@ -356,205 +372,273 @@ defmodule MLLP.Client do
+    backoff
+    |> :backoff.get()
+    |> :timer.seconds()
   end
 
-  defp maybe_close(state), do: state
+  #########################
+  #### Receiving state ####
+  #########################
+  def receiving(:enter, :connected, _state) do
+    Logger.debug("Waiting for response...")
+    :keep_state_and_data
+  end
 
-  defp recv_ack(state, timeout) do
-    recv_ack(state, {timeout, 0}, <<>>)
+  def receiving({:call, from}, {:send, _message, _options}, _state) do
+    {:keep_state_and_data, [{:reply, from, format_reply({:error, :busy_with_other_call}, :send)}]}
   end
 
-  defp recv_ack(_state, {time_left, time_owed}, _buffer)
-       when is_integer(time_left) and time_left <= time_owed do
-    {:error, :timeout}
+  def receiving(:state_timeout, :receive_timeout, state) do
+    {:next_state, :connected, reply_to_caller({:error, :timeout}, state)}
   end
 
-  defp recv_ack(state, {time_left, time_owed}, buffer) do
-    {res, elapsed} = do_recv(state, 0, time_left)
+  def receiving(:info, {transport, socket, data}, %{socket: socket} = state)
+      when transport in [:tcp, :ssl] do
+    new_data = handle_received(data, state)
+    next_state = (new_data.caller && :receiving) || :connected
+    {:next_state, next_state, new_data}
+  end
 
-    case res do
-      {:ok, reply} ->
-        new_buf = buffer <> reply
-        check = byte_size(new_buf) - 3
+  def receiving(:info, {transport_closed, socket}, %{socket: socket} = state)
+      when transport_closed in [:tcp_closed, :ssl_closed] do
+    {:next_state, :disconnected, handle_closed(state)}
+  end
+
+  def receiving(:info, {transport_error, socket, reason}, %{socket: socket} = state)
+      when transport_error in [:tcp_error, :ssl_error] do
+    {:next_state, :disconnected, handle_error(reason, maybe_close(reason, state))}
+  end
 
-        case new_buf do
-          <<@header, _ack::binary-size(check), @trailer>> ->
-            {:ok, new_buf}
+  def receiving(event, unknown, _state) do
+    unexpected_message(:receiving, event, unknown)
+  end
 
-          <<@header, _rest::binary>> ->
-            time_credit = update_recv_time_credit(time_left, time_owed + elapsed)
-            recv_ack(state, time_credit, new_buf)
+  ########################################
+  ### End of :gen_statem callbacks ###
+  ########################################
 
-          _ ->
-            {:error, :invalid_reply}
-        end
+  defp unexpected_message(state, event, message) do
+    Logger.warn(
+      "Event: #{inspect(event)} in state #{state}. Unknown message received => #{inspect(message)}"
+    )
 
-      {:error, _} = err ->
-        err
+    :keep_state_and_data
+  end
+
+  ## Handle the (fragmented) responses to `send` request from a caller
+
+  defp handle_received(_reply, %{caller: nil} = state) do
+    ## No caller, ignore
+    state
+  end
+
+  defp handle_received(reply, %{receive_buffer: buffer} = state) do
+    new_buf = (buffer && buffer <> reply) || reply
```

Comment:
Do you think we should switch to an iolist here? It's an optimization, but yeah.

Reply:
:+1: for io lists, even if it optimization doesn't matter, I'd say it is semantically better anyway. The buffer can always be initialized to an empty list as well, so this line becomes:

```suggestion
    new_buf = [buffer | reply]
```

PS: you could also initialize it to an empty string and always concatenate.

Reply:
Indeed. Maybe commit the suggest or push the the rest? Or similar. I think that was implicit per the suggestion IMHO.

Reply:
`buffer <> reply` will copy both sides on every operation, given the buffer increases, the amount being copied is ever increasing. So it is more costly in terms of CPU and memory. If you do `[buffer | reply]` and at the end you do `IO.iodata_to_binary(buffer)`, then everything is just copied once (with no increasing costs).

Reply:
This may be a premature comment, but when it comes to checking if we have everything (i.e, a full ack) what we can do is something like : 

```
check = byte_size(new_fragment) - 2 
case {receiver_buffer, new_fragment]} do 
  {[^@header | _],  <<_::binary-size(check), @trailer>>]  -> Enum.join(receiver_buffer)
  _ -> :moar_data_coming
 end 
```

Something like that, this allows us to defer the concatenation until the very end 😄

Reply:
haha, I think we had a race condition on the suggestion here! 😄 Mine is flawed in that I suggested Enum.join, and IO.iodata_to_binary/1 is going to be much better here 😁

Reply:
Sorry for the noise, but there's one other check you have to make, it's an edge case. 

```
case {receiver_buffer, new_fragment]} do 
   {[^@header | _],  <<_::binary-size(check), @trailer>>]}  -> IO.iodata_to_binary(receiver_buffer)
    {[^@header | _],  <<_::binary-size(check - 1), @first_part_of_trailer::binary-size(1)>>, @carriage_return]}  -> IO.iodata_to_binary(receiver_buffer)
    _ -> :moar_data_coming
 end
```

There might be a nicer way to do that, but I'm not aware of one.

Reply:
> `buffer <> reply` will copy both sides on every operation, given the buffer increases, the amount being copied is ever increasing. So it is more costly in terms of CPU and memory. If you do `[buffer | reply]` and at the end you do `IO.iodata_to_binary(buffer)`, then everything is just copied once (with no increasing costs).


Thanks, @josevalim
This makes sense. 
We will have to rewrite the code then to check every incoming chunk of data for the trailer being in there **_before_** we do [buffer | reply]. Which is what I am going to do :-)